clang/tools/offload-arch/LevelZeroArch.cpp

```diff
@@ -0,0 +1,128 @@
+
+#define DEFINE_WRAPPER(NAME)                                                   \
+  using NAME##_ty = decltype(NAME);                                            \
+  void *NAME##Ptr = nullptr;                                                   \
+  template <class... Ts> ze_result_t NAME##_wrap(Ts... args) {                 \
+    if (!NAME##Ptr) {                                                          \
+      return ZE_RESULT_ERROR_UNKNOWN;                                          \
+    }                                                                          \
+    return reinterpret_cast<NAME##_ty *>(NAME##Ptr)(args...);                  \
+  };
+
+DEFINE_WRAPPER(zeInitDrivers)
+DEFINE_WRAPPER(zeDeviceGet)
+DEFINE_WRAPPER(zeDeviceGetProperties)
+
+static bool loadLevelZero() {
+  const char *L0Library = "libze_loader.so";
+  std::string ErrMsg;
+
+  auto DynlibHandle = std::make_unique<llvm::sys::DynamicLibrary>(
+      llvm::sys::DynamicLibrary::getPermanentLibrary(L0Library, &ErrMsg));
+  if (!DynlibHandle->isValid()) {
+    if (ErrMsg.empty())
+      ErrMsg = "unknown error";
+    if (Verbose)
+      llvm::errs() << "Unable to load library '" << L0Library << "': " << ErrMsg
+                   << "\n";
+    return false;
+  }
+
+  constexpr struct {
+    const char *name;
+    void **fptr;
+  } dlwrap[] = {
+      {"zeInitDrivers", &zeInitDriversPtr},
+      {"zeDeviceGet", &zeDeviceGetPtr},
+      {"zeDeviceGetProperties", &zeDeviceGetPropertiesPtr},
+  };
+
+  for (auto entry : dlwrap) {
+    void *P = DynlibHandle->getAddressOfSymbol(entry.name);
+    if (P == nullptr) {
+      if (Verbose)
+        llvm::errs() << "Unable to find '" << entry.name << "' in '"
+                     << L0Library << "'\n";
+      return false;
+    }
+    *(entry.fptr) = P;
+  }
+
+  return true;
+}
+
+#define CALL_ZE_AND_CHECK(Fn, ...)                                             \
+  do {                                                                         \
+    ze_result_t Rc = Fn##_wrap(__VA_ARGS__);                                   \
+    if (Rc != ZE_RESULT_SUCCESS) {                                             \
+      if (Verbose)                                                             \
+        llvm::errs() << "Error: " << __func__ << ":" << #Fn                    \
+                     << " failed with error code " << Rc << "\n";              \
+      return 1;                                                                \
+    }                                                                          \
+  } while (0)
+
+int printGPUsByLevelZero() {
+  if (!loadLevelZero())
+    return 1;
+
+  ze_init_driver_type_desc_t driver_type = {};
+  driver_type.stype = ZE_STRUCTURE_TYPE_INIT_DRIVER_TYPE_DESC;
+  driver_type.flags = ZE_INIT_DRIVER_TYPE_FLAG_GPU;
+  driver_type.pNext = nullptr;
+  uint32_t driverCount{0};
+
+  // Initialize and find all drivers
+  CALL_ZE_AND_CHECK(zeInitDrivers, &driverCount, nullptr, &driver_type);
```

Comment:
do we need to do this twice? reading the spec for the function i didnt immediately see why we would need to

Reply:
Yes, the first time you get the count. With that you allocate the proper storage and call it again to get the actual data.

Reply:
ah ok sorry we do it just for the vector size, got it